File: backend/app/models/endpoints.py
"""`endpoints` and `endpoint_models` — an endpoint and every model seen on it."""

from datetime import datetime
from typing import Literal

from sqlalchemy import ForeignKey, Text, UniqueConstraint, false, func
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base

#: How a model came to be recorded on an endpoint.
EndpointModelSource = Literal["discovered", "manual", "run"]


class Endpoint(Base):
    """An OpenAI-compatible endpoint: a base URL, an optional API key, and
    free-text hardware notes.

    Anything that speaks the protocol is one — an Ollama or vLLM box you own, a
    proxy, or a hosted frontier API. The hardware notes are half the product's
    answer where the endpoint *is* a box ("which model is good enough" is worth
    little without "and what box does it need"), which is why every run names
    the endpoint that produced its numbers, and why they are optional.

    `base_url` / `api_key` are credentials and are therefore read **live** at
    execution time rather than frozen into a run, so a moved endpoint does not
    break Resume.
    """

    __tablename__ = "endpoints"

    id: Mapped[int] = mapped_column(primary_key=True)
    customer_id: Mapped[int] = mapped_column(
        ForeignKey("customers.id", ondelete="RESTRICT"), index=True
    )
    #: Readable from **every** workspace, writable only from the one that owns
    #: it — which is why it may only be set on a row owned by the Base
    #: workspace (`app.repos.endpoints` enforces that on create and on update,
    #: so no call site can forget). A consultancy runs the same DGX Spark
    #: across every engagement, and re-registering it per workspace duplicates
    #: an API key and guarantees half the copies go stale.
    #:
    #: The sharing is expressed as one read-side seam,
    #: `app.scope.visible_where` — `customer_id` stays the ownership predicate
    #: every UPDATE and DELETE uses, so "read-only outside Base" needs no
    #: permission layer.
    is_global: Mapped[bool] = mapped_column(server_default=false())
    name: Mapped[str]
    base_url: Mapped[str]
    api_key: Mapped[str | None]
    cpu: Mapped[str | None]
    ram: Mapped[str | None]
    gpu: Mapped[str | None]
    notes: Mapped[str | None]
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())


class EndpointModel(Base):
    """Every model ever seen on an endpoint. Never deleted from.

    Discovery upserts and flips `currently_loaded` false for models absent from
    `/v1/models`; manual adds and run creation upsert too (`source`). Keeping
    the history is what lets a past run still name a model the endpoint no
    longer serves.

    Scope is inherited through `endpoint_id`; there is no `customer_id` here.

    On a **global** endpoint that means the history accumulates across every
    engagement: discovery, a manual add and every run write here, and on a
    shared box those runs come from whichever workspace booked it. That is
    intended, not a leak — shared hardware has one shared history, and "this
    box has already served qwen3:32b" is exactly what the next engagement needs
    to know. Nothing customer-specific lands here; a row is a model id, a
    timestamp and how it was first seen.
    """

    __tablename__ = "endpoint_models"

    id: Mapped[int] = mapped_column(primary_key=True)
    endpoint_id: Mapped[int] = mapped_column(ForeignKey("endpoints.id", ondelete="CASCADE"))
    model_id: Mapped[str]
    currently_loaded: Mapped[bool] = mapped_column(server_default=false())
    first_seen_at: Mapped[datetime] = mapped_column(server_default=func.now())
    last_seen_at: Mapped[datetime] = mapped_column(server_default=func.now())
    source: Mapped[EndpointModelSource] = mapped_column(Text)

    __table_args__ = (UniqueConstraint("endpoint_id", "model_id"),)

File: frontend/src/api/endpoints.ts
// The routes this module talks to (backend/app/api/endpoints.py):
//
//   GET    /api/endpoints                 -> Endpoint[]  (counts embedded, the
//                                            list page's "N/M loaded" column)
//   POST   /api/endpoints                  EndpointInput -> Endpoint
//   GET    /api/endpoints/{id}            -> Endpoint
//   PUT    /api/endpoints/{id}             EndpointInput -> Endpoint
//   DELETE /api/endpoints/{id}            -> (204)
//
// `api_key` is write-only: the response never carries it, only `has_api_key`.
// On `PUT` it is the one field that is *not* a full replacement — omit it and
// the stored key is left alone, send `null`/`""` to clear it deliberately, send
// a value to replace it. So a caller must not spread a form's empty string into
// the body, or every save wipes the key.
//   GET    /api/endpoints/{id}/models     -> EndpointModel[]  (loaded first)
//   POST   /api/endpoints/{id}/models      { model_id } -> EndpointModel  (manual
//                                            add; an upsert, so a model already
//                                            on the endpoint is answered with as
//                                            it stands rather than refused)
//   POST   /api/endpoints/{id}/discover   -> DiscoverModelsResult
//   POST   /api/endpoints/{id}/test       -> TestConnectionResult
//
// Discover/test answer 200 with a discriminated `ok` union rather than an
// HTTP error status, because an unreachable endpoint is an expected outcome
// of the probe itself, not a failure of the request to this API — same
// distinction the old Next.js buttons made.
import { api } from './client'

export interface Endpoint {
  id: number
  name: string
  base_url: string
  /** Whether a key is stored — the key itself never leaves the server. */
  has_api_key: boolean
  cpu: string | null
  ram: string | null
  gpu: string | null
  notes: string | null
  /** Shared with every workspace by the Base workspace that owns it. */
  is_global: boolean
  /** Whether *this* workspace owns the row — false only for a global endpoint
   * seen from elsewhere, which is exactly when edit/delete/API-key controls
   * are hidden rather than rendered disabled. */
  editable: boolean
  created_at: string
  updated_at: string
  model_count: number
  loaded_model_count: number
}

export interface EndpointInput {
  name: string
  base_url: string
  /** Omit to keep the stored key, `null` to clear it, a string to replace it. */
  api_key?: string | null
  cpu?: string | null
  ram?: string | null
  gpu?: string | null
  notes?: string | null
  /** Refused by the API outside the Base workspace. Patch-like on `PUT`: omit
   * to leave the stored flag alone, since it defaults to `false` and a caller
   * that does not know about sharing must not un-share the row on every save. */
  is_global?: boolean
}

export type EndpointModelSource = 'discovered' | 'manual' | 'run'

export interface EndpointModel {
  id: number
  endpoint_id: number
  model_id: string
  currently_loaded: boolean
  first_seen_at: string
  last_seen_at: string
  source: EndpointModelSource
}

export type DiscoverModelsResult =
  | { ok: true; discovered: number; retired: number; models: string[] }
  | { ok: false; error: string }

export type TestConnectionResult =
  | { ok: true; status: number; latency_ms: number }
  | { ok: false; error: string; status?: number }

export const endpointsApi = {
  list: () => api.get<Endpoint[]>('/endpoints'),
  get: (id: number) => api.get<Endpoint>(`/endpoints/${id}`),
  create: (input: EndpointInput) => api.post<Endpoint>('/endpoints', input),
  update: (id: number, input: EndpointInput) => api.put<Endpoint>(`/endpoints/${id}`, input),
  remove: (id: number) => api.delete<void>(`/endpoints/${id}`),
  listModels: (id: number) => api.get<EndpointModel[]>(`/endpoints/${id}/models`),
  addModel: (id: number, modelId: string) =>
    api.post<EndpointModel>(`/endpoints/${id}/models`, { model_id: modelId }),
  discover: (id: number) => api.post<DiscoverModelsResult>(`/endpoints/${id}/discover`),
  test: (id: number) => api.post<TestConnectionResult>(`/endpoints/${id}/test`),
}
